ambition_utils/rrule/models.py

```diff
@@ -330,16 +330,20 @@ def generate_dates(self, num_dates=20):
             # Convert to time zone
             date_with_tz = fleming.attach_tz_if_none(d, self.time_zone)
             date_in_utc = fleming.convert_to_tz(date_with_tz, pytz.utc, True)
-            dates.append(date_in_utc)
-
-            for x in range(0, num_dates):
+            
+            if not start_date or date_in_utc > start_date:
+                dates.append(date_in_utc)
+            
+            while len(dates) < num_dates:
```

Comment:
is there any chance this can infinite loop for example if trying to generate 20 dates, but the rrule definition stops before 20 dates? the previous logic would generate up to a certain number of dates based on the rrule definition, but stop once the `.after` call returns null.

we have a method to force date generation by using the RRule's `get_next_occurrence` with `force=True`. maybe rather than using `rule.after` down below, we use `self.get_next_occurrence` while passing force=True and last_occurrence.

if we need to force all dates, then `get_dates` needs an additional arg to handle the `force` option. this method's intent is to show a preview of the upcoming dates, but not go beyond the actual rrule definition. if we don't need a force option here, then we just need to restore the original logic to break out of the `while` loop if the next date is null.

Reply:
This was a great find.

The logic is not using `get_next_occurrence` in the loop. (I explored using it outside of the loop as well but it doesn't work as needed if there is no `last_occurrence` set for a new object. 

There is also not a need for `num_dates` to trump `start_date` so nothing is forced in the loop either. 

I added `test_get_dates_num_dates_greater` to flex. I also simplified some datetime used in tests by dropping the hour. They weren't necessary and only added (minor) confusion.